Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -484,25 +546,55 @@ public CsharpCompiler(ErrorReporter reporter)
       }
     }
 
+    public override bool NeedsCustomReceiver(MemberDecl member) {
+      Contract.Requires(member != null);
+      //Dafny and C# have different ideas about variance, so not every datatype member can be in the interface.
+      if (!member.IsStatic && (member.EnclosingClass is IndDatatypeDecl || member.EnclosingClass is CoDatatypeDecl)) {
+        DatatypeDecl d = member.EnclosingClass is IndDatatypeDecl ind ? ind : member.EnclosingClass as CoDatatypeDecl;
+        foreach (var tp in d.TypeArgs) {
+          Predicate<Type> InvalidType = ty => ty.AsTypeParameter != null && ty.AsTypeParameter.Equals(tp)
+                                              || ty.TypeArgs.Exists(a => a.AsTypeParameter.Equals(tp));
```

Comment:
I can never remember which of `||` and `&&` binds tighter, so I always add parentheses.

Reply:
This finds type `TP` and `seq<TP>`, but what about `seq<seq<TP>>` or other more deeply nested occurrences of `TP`? What you want is something like `FreeTypeVariables(ty).Contains(tp)`.

Reply:
Or, what would you think of simplifying the design and replacing L555-L571 with just `return true;`?